로그를 이용한 비교하기 구현

qkrdmstlr3 opened the pull request:
## 상세사항
- 로그 선택시 그 상태의 로그와 비교하는 기능 추가
- 히스토리 모달 창 등 비교하기를 위한 지도 관련 스타일 조정
<img width="1410" alt="스크린샷 2020-12-08 오후 5 38 14" src="https://user-images.githubusercontent.com/26402298/101460016-5c2b4680-397c-11eb-9a83-72cc109b516d.png">

qkrdmstlr3 committed:
[refactor] getCompare함수에서 mousemove제거

- 가운데에 움직일 수 있는 버튼이 있음으로 mousemove를 제거함
- css호출로 버튼을 사용가능하도록 만들어줌

qkrdmstlr3 committed:
[refactor] [#77] Map의 css조절

- map의 css의 position등 속성이 잘못되어 있어서, compare시 양 옆지도의 싱크가 맞지않음
- position속성을 가장 바깥쪽 div에 주는 형식으로 해결

qkrdmstlr3 committed:
[refactor] [#77] 지도 버튼들의 위치 조정 및 History component 스타일 조정

- compare시 먹히는 현상이 있어서 지도 버튼들의 위치를 밖으로 빼내었다
- 따라서 History컴포넌트의 스타일링을 조금 수정하였다

qkrdmstlr3 committed:
[refactor] Map / History컴포넌트 스타일링

- map / history컴포넌트를 추가적으로 스타일링
- 로그를 이쪽에서 남긴다

qkrdmstlr3 committed:
[refactor] [#77] history reducer수정

- 로그를 추가하는 부분을 중복과 display를 제거

qkrdmstlr3 committed:
[feat] 로그 클릭시 그 때의 상태와 비교하는 기능 추가

- 로그를 클릭하면 그 상태와 현재 상태가 반으로 갈라져서 보여진다
- setFeatureStyle를 이용해서 지도 로딩 후 스타일을 변경시켜주었다

sayhs30 commented on src/components/Map/History/History.tsx:
```diff
@@ -63,19 +74,27 @@ function History({
+      <HistoryList>
+        {log ? (
+          log
+            .map((item) => (
```
이 구문은 (log || []).map ~ 형태로 줄여도 될거같네요 ㅎ

qkrdmstlr3 replied:
아 그렇네요. null이나 빈 값들이 들어가는 곳에 있어서는 항상 유념해야겠습니다!

qkrdmstlr3 replied:
아 그렇네요. null이나 빈 값들이 들어가는 곳에 있어서는 항상 유념해야겠습니다!

sayhs30 commented on src/components/Map/History/History.tsx:
```diff
@@ -63,19 +74,27 @@ function History({
+            .map((item) => (
+              <HistoryItem
+                key={item.id}
+                onClick={() => comparisonButtonClickHandler(item.id as string)}
```
불필요한 렌더링을 줄이기 위해 useCallback 함수를 이용하는건 어떨까요!?

sayhs30 commented on src/components/Map/History/History.tsx:
```diff
@@ -63,19 +74,27 @@ function History({
+                key={item.id}
+                onClick={() => comparisonButtonClickHandler(item.id as string)}
+              >
+                <p>{`${item.feature} > ${item.subFeature} > ${item.element} > ${item.subElement}`}</p>
```
별도 함수로 빼서, 이름/의미를 부여한다면 좀더 가독성을 높일수있을거같습니다.

sayhs30 commented on src/hooks/map/useCompareFeature.ts:
```diff
@@ -41,23 +47,42 @@ function useCompareFeature({
 
+    beforeMap.on('load', () => {
+      if (!history) return;
+      const [log] = history.log?.filter((log) => log.id === logId) || [];
```
아래 if문에서 return되면, 이 코드가 실행되는 의미가 없네요. 위치를 이동하는게 좋을거같네요 ㅎ

sayhs30 commented on src/store/history/reducer.ts:
```diff
@@ -32,8 +18,8 @@ function historyReducer(
     case INIT_HISTORY: {
-      const log = localStorage.getItem('log')
-        ? JSON.parse(localStorage.getItem('log') as string)
+      const log = localStorage.getItem(logKey)
```
localStorage 를 이용한 비슷한 동작이 반복되는데, 별도로 캡슐화해서 관리하는것도 좋을거같네요~

qkrdmstlr3 replied:
흠 그렇네요... 일단 localStorage와의 상호작용을 최소화하자고 팀 내에서 이야기가 나왔는데, 참고해서 구현하겠습니다!

sayhs30 commented on src/store/history/reducer.ts:
```diff
@@ -56,31 +42,25 @@ function historyReducer(
       const id = getRandomId(8);
       const newState = JSON.parse(JSON.stringify(state));
```
다른 reducer 코드에서도 JSON.parse(JSON.stringify(state)) 이 코드가 많은데, 
new state를 생성하기 위해서 인건가요?
만약 맞다면 immer.js와 같은 immutable 다루는 라이브러리도 있는데, 기회되면 보시면 도움될거같습니다~!

qkrdmstlr3 replied:
일단 deepcopy가 필요한 부분은 저렇게 처리를 하고 다음 주 리팩토링 시 deepcopy함수를 만들어서 집어넣자고 이야기가 나왔습니다!

Eunsol0410 replied:
이슈에 해당 내용 넣어둘게요!

ghost commented:
멋진데요! 고생하셨습니다 ㅎㅎ 

보다보니 궁금증이 생겼는데요, 혹시 히스토리 리스트 아이템 선택 시 지도 스타일 이외 UI 설정 창들의 값도 함께 변경이 되나요?ㅎ

qkrdmstlr3 commented:
그렇지는 않습니다! 메인화면 기준이라서, 비교되는 지도(좌측 지도)는 보여주는 기능만 합니다. 혹시 비교되는 화면에 대한 것을 옆에 같이 표기를 해주게 된다면, 사용자 입장에서 좀 더 느껴지는 점이 있을까요??

ghost commented:
> 그렇지는 않습니다! 메인화면 기준이라서, 비교되는 지도(좌측 지도)는 보여주는 기능만 합니다. 혹시 비교되는 화면에 대한 것을 옆에 같이 표기를 해주게 된다면, 사용자 입장에서 좀 더 느껴지는 점이 있을까요??

아 제가 질문드린건, 비교기능이 아니라, 히스토리 리스트가 있길래 이쪽 상태관리는 어떻게하셨나 궁금해서 여쭤본거예요~ 
본 PR의 주제에서 벗어난 질문입니다 ㅎ

qkrdmstlr3 commented:
아 비교시에는 각 로그에다가 그 때의 전체상태를 저장해 놓은 후 꺼내서 스타일을 적용하는 방식으로 하고있습니다. 여담으로 이전상태에서의 변경사항을 같이 저장하기 때문에, 뒤로/앞으로 가기시에는 전체를 바꾸기 보다는 차이를 이용해서 좀 더 효율적으로 구현하고자 하였습니다!

qkrdmstlr3 committed:
[feat] Overlay컴포넌트 추가

- overlay컴포넌트를 추가하고 기존 모달들에게 달아주었다.

qkrdmstlr3 committed:
[refactor] 리뷰 내용 반영

- 리뷰 내용을 반영하여서 코드를 좀더 간결하게 줄임
- history목록을 함수로 빼는 내용은 추후 스타일링을 하면서 진행하면 될듯
- 👍

qkrdmstlr3 committed:
[feat] 테마 기능 동작하게 함 + 청사진 테마 연동

- 테마컴포넌트에 hook을 달아서 테마를 클릭할 시 useWholeStyle hook을 이용해서 스타일을 변경시킬 수 있도록 함
- 청사진 테마로 테스트

qkrdmstlr3 committed:
[feat] poi초기화 json파일 작성

- poi를 초기화 할 수 있는 json파일을 작성하였다

qkrdmstlr3 commented on src/utils/rendering-data/layers/poi.json:
```diff
@@ -0,0 +1,889 @@
+{
```
준희님 아마 정훈님이 올려주신 자료에있는 json 레이어의 형태를 참고해서 store/map/layers에 있는 네이버 레이어들 + mapbox레이어에 대한 json파일을 만드시면 되요. map-styling/transit 보시면 이전에 정훈님이 정리해 놓으신 자료있어요. 레이어들의 색상은 rendering-data폴더의 initLayersColor와 layersColor참고하시면 되요!

qkrdmstlr3 committed:
[refactor] ExportModal스타일링

- exportmodal에 대한 스타일링을 진행하였다

qkrdmstlr3 committed:
Merge branch 'feature/history' into feature/history
